File: smoke_tests/swift/insecure_random.swift
// Insecure Randomness vulnerabilities in Swift
import Foundation
import Security

class InsecureRandom {

    // Test 1: arc4random for security token (older systems)
    func generateTokenArc4() -> String {
        // VULNERABLE: arc4random may not be crypto-secure on all platforms
        var token = ""
        for _ in 0..<32 {
            token += String(format: "%02x", arc4random_uniform(256))
        }
        return token
    }

    // Test 2: drand48 for security purposes
    func generateTokenDrand48() -> Double {
        // VULNERABLE: drand48 is not cryptographically secure
        srand48(Int(Date().timeIntervalSince1970))
        return drand48()
    }

    // Test 3: random() for security
    func generateWeakRandom() -> Int {
        // VULNERABLE: random() is predictable
        srandom(UInt32(time(nil)))
        return random()
    }

    // Test 4: Time-based seed
    func generateTimeSeeded() -> UInt32 {
        // VULNERABLE: Time-based seed is predictable
        srand(UInt32(time(nil)))
        return UInt32(rand())
    }

    // Test 5: UUID for secrets
    func generateUUIDToken() -> String {
        // VULNERABLE: UUID v4 may not be crypto random on all implementations
        return UUID().uuidString
    }

    // Test 6: Int.random for crypto
    func generateIntRandom() -> Int {
        // Note: Swift's Int.random uses a secure generator, but pattern is risky
        // VULNERABLE: Depending on usage context
        return Int.random(in: 0..<1000000)
    }

    // Test 7: Array shuffling for security
    func shuffleForSecurity<T>(array: [T]) -> [T] {
        // VULNERABLE: shuffle() may not use crypto random on all platforms
        var mutableArray = array
        mutableArray.shuffle()
        return mutableArray
    }

    // Test 8: rand() from C library
    func cRand() -> Int32 {
        // VULNERABLE: C rand() is not secure
        return rand()
    }

    // Test 9: Predictable session ID
    func generateSessionId(userId: Int) -> String {
        // VULNERABLE: Predictable session ID
        let timestamp = Int(Date().timeIntervalSince1970)
        return "\(userId)-\(timestamp)"
    }

    // Test 10: Weak OTP generation
    func generateOTP() -> String {
        // VULNERABLE: Using non-crypto random for OTP
        let otp = arc4random_uniform(1000000)
        return String(format: "%06d", otp)
    }

    // Test 11: Password generation with weak random
    func generatePassword(length: Int) -> String {
        let chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        var password = ""
        // VULNERABLE: Not guaranteed crypto-secure on all platforms
        for _ in 0..<length {
            let index = Int(arc4random_uniform(UInt32(chars.count)))
            password += String(chars[chars.index(chars.startIndex, offsetBy: index)])
        }
        return password
    }

    // Test 12: IV generation with weak random
    func generateIV() -> Data {
        // VULNERABLE: IV should use SecRandomCopyBytes
        var iv = Data(count: 16)
        for i in 0..<16 {
            iv[i] = UInt8(arc4random_uniform(256))
        }
        return iv
    }

    // Test 13: Salt generation with weak random
    func generateSalt() -> Data {
        // VULNERABLE: Salt should use crypto random
        var salt = Data(count: 32)
        for i in 0..<32 {
            salt[i] = UInt8(arc4random_uniform(256))
        }
        return salt
    }

    // Secure alternative (for reference):
    func secureRandom(count: Int) -> Data? {
        var bytes = [UInt8](repeating: 0, count: count)
        let status = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
        return status == errSecSuccess ? Data(bytes) : nil
    }
}

File: smoke_tests/lua/insecure_random.lua
-- Insecure Random Number Generation in Lua

-- Test 1: math.random without proper seed
function generate_token_weak()
    -- VULNERABLE: Default seed is predictable
    local token = ""
    for i = 1, 32 do
        token = token .. string.char(math.random(65, 90))
    end
    return token
end

-- Test 2: Time-based seed
function generate_session_id()
    -- VULNERABLE: Time-based seed is predictable
    math.randomseed(os.time())
    return math.random(1000000, 9999999)
end

-- Test 3: Weak password generation
function generate_password(length)
    math.randomseed(os.time())
    local chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    local password = ""
    -- VULNERABLE: Predictable random
    for i = 1, length do
        local idx = math.random(1, #chars)
        password = password .. chars:sub(idx, idx)
    end
    return password
end

-- Test 4: UUID generation
function generate_uuid()
    math.randomseed(os.time())
    -- VULNERABLE: Predictable UUID
    local template = "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx"
    return template:gsub("[xy]", function(c)
        local v = (c == "x") and math.random(0, 15) or math.random(8, 11)
        return string.format("%x", v)
    end)
end

-- Test 5: OTP generation
function generate_otp()
    math.randomseed(os.time())
    -- VULNERABLE: Predictable OTP
    return string.format("%06d", math.random(0, 999999))
end

-- Test 6: API key generation
function generate_api_key()
    math.randomseed(os.time())
    -- VULNERABLE: Predictable API key
    local key = ""
    for i = 1, 32 do
        key = key .. string.format("%02x", math.random(0, 255))
    end
    return key
end

-- Test 7: CSRF token generation
function generate_csrf_token()
    math.randomseed(os.time())
    -- VULNERABLE: Predictable CSRF token
    return tostring(math.random(10000000, 99999999))
end

-- Test 8: Nonce generation
function generate_nonce()
    math.randomseed(os.time())
    -- VULNERABLE: Predictable nonce
    local nonce = ""
    for i = 1, 16 do
        nonce = nonce .. string.format("%02x", math.random(0, 255))
    end
    return nonce
end

-- Test 9: Encryption IV generation
function generate_iv()
    math.randomseed(os.time())
    -- VULNERABLE: Predictable IV
    local iv = ""
    for i = 1, 16 do
        iv = iv .. string.char(math.random(0, 255))
    end
    return iv
end

-- Test 10: Salt generation
function generate_salt()
    math.randomseed(os.time())
    -- VULNERABLE: Predictable salt
    local salt = ""
    for i = 1, 16 do
        salt = salt .. string.format("%02x", math.random(0, 255))
    end
    return salt
end

-- Test 11: Shuffle with weak random
function shuffle_array(arr)
    math.randomseed(os.time())
    -- VULNERABLE: Predictable shuffle
    for i = #arr, 2, -1 do
        local j = math.random(i)
        arr[i], arr[j] = arr[j], arr[i]
    end
    return arr
end

-- Test 12: Random selection
function select_random_winner(participants)
    math.randomseed(os.time())
    -- VULNERABLE: Predictable winner selection
    local idx = math.random(1, #participants)
    return participants[idx]
end

-- Test 13: Process ID as seed (slightly better but still weak)
function generate_token_pid_seed()
    -- VULNERABLE: PID + time still predictable
    math.randomseed(os.time() + (os.getenv("PID") or 0))
    return math.random(1000000000, 9999999999)
end
